### PR TITLE
KAFKA-7483: Allow streams to pass headers through Serializer.

### PR DESCRIPTION
**Pull request by kamalcph**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by kamalcph** [comment]
@vvcephei @mjsax 
Could you please review this PR ?

File: streams/src/test/java/org/apache/kafka/streams/state/KeyValueStoreTestDriver.java
Comment (kamalcph) [inline streams/src/test/java/org/apache/kafka/streams/state/KeyValueStoreTestDriver.java:210]: Please double check the methods `keyFrom`, `valueFrom`, `rawKey` and `rawValue` in StateSerdes.java. I think no change is required there.

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/state/KeyValueStoreTestDriver.java:210]: `StateSerdes` is use to read/write into local stores and changlog topics. Because changelog topic should only be read/written by Kafka Stream, it seems we don't need a change there to me.

**Review by mjsax — APPROVED** [review]
LGMT, assuming Jenkin passes.

**Comment by kamalcph** [comment]
retest this please

**Review by satishd — APPROVED** [review]
Thanks @kamalcph for the PR.
+1 LGTM

**Comment by mjsax** [comment]
Call for second review (anyof) @guozhangwang @bbejeck @vvcephei 

Retest this please

**Comment by bbejeck** [comment]
retest this please

**Review by bbejeck — APPROVED** [review]
Thanks for the PR @kamalcph, LGTM

File: streams/src/test/java/org/apache/kafka/streams/state/KeyValueStoreTestDriver.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/state/KeyValueStoreTestDriver.java:210]: @mjsax is right. Just to clarify, state store / changelogs today only do header-agnostic serde, so the scope of this PR is only for sink nodes.

**Comment by guozhangwang** [comment]
LGTM, thanks @kamalcph ! I will merge it once jenkins passed.

**Comment by guozhangwang** [comment]
Cherry-picked to 2.1

**Pull request merged**
